OCPBUGS-49291:  Improving the DevExp for passing the CSP directives to console per flag + Make use of connect-src and object-src directives

jhadvig opened the pull request:
Improving the DevExp for passing the CSP directives to console per flag instead of JSON.
New usage will be:
```
./bin/bridge --content-security-policy ScriptSrc="localhost:1234 localhost:2345" --content-security-policy FontSrc="localhost:3456 localhost:4567"
```

Also adding connect-src and object-src directives. These will be set to `'none'` if no plugin will specify `source-expression-list` for them. This API was added by https://issues.redhat.com/browse/OCPBUGS-48740

/assign @TheRealJon @spadgett

spadgett commented on pkg/utils/utils.go:
```diff
@@ -109,6 +118,16 @@ func BuildCSPDirectives(k8sMode, pluginsCSP, indexPageScriptNonce string) ([]str
 	scriptSrcDirective = append(scriptSrcDirective, []string{unsafeEval, nonce}...)
 	styleSrcDirective = append(styleSrcDirective, unsafeInline)
 
+	// If the objectSrc directive is not set, add 'none' to it.
+	if len(objectSrcDirective) == 1 && objectSrcDirective[0] == "object-src" {
+		objectSrcDirective = append(objectSrcDirective, none)
+	}
+
+	// If the connectSrc directive is not set, add 'none' to it.
+	if len(connectSrcDirective) == 1 && connectSrcDirective[0] == "connect-src" {
+		connectSrcDirective = append(connectSrcDirective, none)
+	}
```
I think specifying `none` overrides `default-src` so things like `self` would no longer work.

jhadvig replied:
ok, I see, in that case we should only set `self`, for both object-src` and `connect-src`

spadgett commented on pkg/utils/utils.go:
```diff
@@ -109,6 +118,16 @@ func BuildCSPDirectives(k8sMode, pluginsCSP, indexPageScriptNonce string) ([]str
+
+	// If the connectSrc directive is not set, add 'none' to it.
+	if len(connectSrcDirective) == 1 && connectSrcDirective[0] == "connect-src" {
+		connectSrcDirective = append(connectSrcDirective, none)
```
For `connect-src`, we want `self` and also a directive for console.redhat.com

spadgett commented on cmd/bridge/main.go:
```diff
@@ -125,7 +125,9 @@ func main() {
-	fContentSecurityPolicy := fs.String("content-security-policy", "", "Content security policy for the console. (JSON as string)")
+
+	consoleCSPFlags := serverconfig.MultiKeyValue{}
+	fs.Var(&consoleCSPFlags, "content-security-policy", "List of CSP directives that are enabled for the console. Each entry consist of csp-directive-name as a key and csp-directive-value as a value.")
```
nit: An example here would be nice (like we provide for i18n-namespaces above)

spadgett commented on README.md:
````diff
@@ -465,6 +465,22 @@ this way, then 'none' will be used. Additionally, violation reporting is throttl
+Example:
+
+```
+./bin/bridge --content-security-policy ScriptSrc="localhost:1234 localhost:2345" --content-security-policy FontSrc="localhost:3456 localhost:4567"
````
Here I would probably the directive name `font-src` instead of `FontSrc`

```suggestion
./bin/bridge --content-security-policy ScriptSrc="localhost:1234 localhost:2345" --content-security-policy font-src="localhost:3456 localhost:4567"
```

jhadvig replied:
I was thinking about it, but then we would need to change the `console-config.yaml` representation, or accept both formats `FontSrc` & `font-src`. Here I would rather go with the first option.

spadgett replied:
I don't know if we need to change console-config.yaml. Can we just translate between the names in config.go instead?

I wouldn't accept both formats. Let's pick one and go with it. If it's difficult, we can stick with PascalCase.

jhadvig replied:
We will need the change to the console-config.yaml representation. Currently the in console-config/yaml CSPs are set in `FontSrc="localhost:3456 localhost:4567"` format. Thats how they are passed to the `--content-security-policy` flag by bridge.  OR we need to update the `addContentSecurityPolicy()` in `pkg/serverconfig/config.go` so the `cspDirectiveName` is transformed into the `font-src` format.

spadgett commented on pkg/utils/utils.go:
```diff
@@ -98,8 +103,12 @@ func BuildCSPDirectives(k8sMode, pluginsCSP, indexPageScriptNonce string) ([]str
+				connectSrcDirective = append(connectSrcDirective, sources...)
 			default:
-				klog.Warningf("ignored invalid CSP directive: %v", directive)
+				klog.Infof("ignored invalid CSP directive: %v", directive)
```
nit: This isn't necessarily an invalid directive since there are many we don't allow to be set

```suggestion
				klog.Infof("ignored unsupported CSP directive: %v", directive)
```

jhadvig commented:
@spadgett I've vendor the `object-src` removal and updated code and tests accordingly.
PTAL

spadgett commented on pkg/utils/utils.go:
```diff
@@ -3,164 +3,122 @@ package utils
+		case connectSrc:
+			directives[connectSrc] = append(directives[connectSrc], sources)
+		default:
+			klog.Infof("ignored unsupported CSP directive: %v", directive)
```
I think we want to fail on unrecognized values.

jhadvig replied:
👍

spadgett commented on pkg/utils/utils_test.go:
```diff
@@ -18,144 +20,140 @@ const (
-		})
-	}
-}
+// func TestParseContentSecurityPolicyConfig(t *testing.T) {
```
If this test is no longer valid, should we just remove it?

spadgett reviewed (commented):
Thanks @jhadvig 👍

jhadvig commented on cmd/bridge/main.go:
```diff
@@ -124,8 +124,10 @@ func main() {
+	fI18NamespacesFlags := fs.String("i18n-namespaces", "", "List of namespaces separated by comma. Example: --i18n-namespaces=plugin__acm,plugin__kubevirt")
+
+	consoleCSPFlags := serverconfig.MultiKeyValue{}
+	fs.Var(&consoleCSPFlags, "content-security-policy", "List of CSP directives that are enabled for the console. Each entry consist of csp-directive-name as a key and csp-directive-value as a value. Example: --content-security-policy ScriptSrc=\"https://example.com\" --content-security-policy FontSrc=\"https://example2.com https://example3.com\"")
```
👍

jhadvig commented:
@spadgett thanks for review. Addressed your comments. PTAL

jhadvig commented:
/retest

spadgett commented on pkg/utils/utils.go:
```diff
@@ -3,164 +3,123 @@ package utils
-		styleSrcDirective = append(styleSrcDirective, httpLocalHost)
+	directives := map[string][]string{
+		baseURI:        {self},
+		defaultSrc:     {self, consoleDot},
```
Can we get away with only adding `consoleDot` to `scriptSrc` and `connectSrc`? It would be better if we can avoid putting it in `defaultSrc`.

spadgett reviewed (commented):
Code looks good. @jhadvig Do you mind cleaning up the git history a little? I don't mind the separate vendor commits, but maybe squash the address comment commits. Thanks!

spadgett commented:
Hmm, looks like there's a build failure anyway

jhadvig commented:
@spadgett squashed last two commits.
The tests are failing due to the vendored changes which are bumping the openshift/api which now requires go-1.23, thats why I've opened:
- https://github.com/openshift/console/pull/14772
- https://github.com/openshift/release/pull/61664

spadgett commented:
/retest

jhadvig commented:
/retest

jhadvig commented:
@spadgett comment addressed. PTAL

jhadvig commented:
/retest

jhadvig commented:
/retest
/cherry-pick release-4.19

spadgett commented on pkg/utils/utils.go:
```diff
@@ -68,38 +70,40 @@ func BuildCSPDirectives(k8sMode, pluginsCSP, indexPageScriptNonce, cspReportingE
+		case connectSrc:
+			connectSrcDirective = append(connectSrcDirective, sources)
+		default:
+			klog.Infof("ignored invalid CSP directive: %v", directive)
```
We should fail here as well

spadgett reviewed (commented):
You have two bump API commits. Not sure if that's intentional.

Thanks @jhadvig

jhadvig commented:
Thank you @spadgett I've addressed your comment and removed unnecessary API bumps. Also added additional commit which is using `klog.Fatalf()`  which is internally calling `os.Exit()`. 
PTAL

spadgett reviewed (approved):
/lgtm

👍 
Thanks @jhadvig
